File: packages/typescript/src/world.ts
export function toPixelInt(r: number, g: number, b: number, a = 255): number {
  return ((a & 255) << 24) | ((r & 255) << 16) | ((g & 255) << 8) | (b & 255);
}

export type RenderTarget = {
  width: number;
  height: number;
  fillRect(x: number, y: number, w: number, h: number, argb: number): void;
};

export class Actor {
  x: number;
  y: number;
  width: number;
  height: number;
  color: number;
  name: string | null = null;
  world: World | null = null;

  constructor(x: number, y: number, width: number, height: number, color: number) {
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
    this.color = color;
  }

  named(name: string): this {
    this.name = name;
    return this;
  }

  tryMove(dx: number, dy: number): boolean {
    if (!this.world) {
      this.x += dx;
      this.y += dy;
      return true;
    }
    const nx = this.x + dx;
    const ny = this.y + dy;
    if (this.world.blocks(nx, ny, this.width, this.height)) return false;
    this.x = nx;
    this.y = ny;
    return true;
  }

  update(_input?: unknown): void {}
}

export class Trigger {
  x: number;
  y: number;
  width: number;
  height: number;
  onEnter: (actor: Actor) => void;
  id: string | null;
  private _inside: Actor[] = [];

  constructor(
    x: number,
    y: number,
    width: number,
    height: number,
    onEnter?: (actor: Actor) => void,
    id: string | null = null
  ) {
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
    this.onEnter = onEnter ?? (() => {});
    this.id = id;
  }

  setOnEnter(fn: (actor: Actor) => void): void {
    const prev = this.onEnter;
    this.onEnter = (a) => {
      prev(a);
      fn(a);
    };
  }

  evaluate(actors: Actor[]): void {
    const now: Actor[] = [];
    for (const actor of actors) {
      if (this._overlaps(actor)) {
        now.push(actor);
        if (!this._inside.includes(actor)) this.onEnter(actor);
      }
    }
    this._inside = now;
  }

  private _overlaps(actor: Actor): boolean {
    return (
      this.x < actor.x + actor.width &&
      this.x + this.width > actor.x &&
      this.y < actor.y + actor.height &&
      this.y + this.height > actor.y
    );
  }
}

export class World {
  static readonly EMPTY = 0;
  static readonly SOLID = 1;

  cols: number;
  rows: number;
  tileSize: number;
  tiles: number[];
  actors: Actor[] = [];
  triggers: Trigger[] = [];
  clearColor: number;
  solidColor: number;

  constructor(cols: number, rows: number, tileSize = 16) {
    this.cols = cols;
    this.rows = rows;
    this.tileSize = tileSize;
    this.tiles = new Array(cols * rows).fill(World.EMPTY);
    this.clearColor = toPixelInt(20, 20, 30);
    this.solidColor = toPixelInt(60, 60, 80);
  }

  static grid(cols: number, rows: number, tileSize = 16): World {
    return new World(cols, rows, tileSize);
  }

  withClearColor(c: number): this {
    this.clearColor = c;
    return this;
  }

  withSolidColor(c: number): this {
    this.solidColor = c;
    return this;
  }

  fillBorder(tile = World.SOLID): void {
    for (let x = 0; x < this.cols; x++) {
      this.setTile(x, 0, tile);
      this.setTile(x, this.rows - 1, tile);
    }
    for (let y = 0; y < this.rows; y++) {
      this.setTile(0, y, tile);
      this.setTile(this.cols - 1, y, tile);
    }
  }

  setTile(tx: number, ty: number, tile: number): void {
    if (!this._inBounds(tx, ty)) throw new Error(`Tile out of bounds (${tx},${ty})`);
    this.tiles[tx + ty * this.cols] = tile;
  }

  getTile(tx: number, ty: number): number {
    if (!this._inBounds(tx, ty)) return World.SOLID;
    return this.tiles[tx + ty * this.cols];
  }

  isSolid(tx: number, ty: number): boolean {
    return this.getTile(tx, ty) === World.SOLID;
  }

  blocks(x: number, y: number, w: number, h: number): boolean {
    const minTx = Math.floor(x / this.tileSize);
    const maxTx = Math.floor((x + w - 1) / this.tileSize);
    const minTy = Math.floor(y / this.tileSize);
    const maxTy = Math.floor((y + h - 1) / this.tileSize);
    for (let tx = minTx; tx <= maxTx; tx++) {
      for (let ty = minTy; ty <= maxTy; ty++) {
        if (this.isSolid(tx, ty)) return true;
      }
    }
    return false;
  }

  add(actor: Actor): void {
    this.actors.push(actor);
    actor.world = this;
  }

  removeActor(name: string): Actor | null {
    const found = this.findActor(name);
    if (found) {
      this.actors = this.actors.filter((a) => a !== found);
      found.world = null;
    }
    return found;
  }

  addTrigger(trigger: Trigger): void {
    this.triggers.push(trigger);
  }

  findActor(name: string): Actor | null {
    return this.actors.find((a) => a.name === name) ?? null;
  }

  findTrigger(id: string): Trigger | null {
    return this.triggers.find((t) => t.id === id) ?? null;
  }

  update(input?: unknown): void {
    for (const a of this.actors) a.update(input);
    for (const t of this.triggers) t.evaluate(this.actors);
  }

  render(ctx: RenderTarget): void {
    ctx.fillRect(0, 0, ctx.width, ctx.height, this.clearColor);
    for (let ty = 0; ty < this.rows; ty++) {
      for (let tx = 0; tx < this.cols; tx++) {
        if (this.isSolid(tx, ty)) {
          ctx.fillRect(tx * this.tileSize, ty * this.tileSize, this.tileSize, this.tileSize, this.solidColor);
        }
      }
    }
    for (const a of this.actors) {
      ctx.fillRect(a.x, a.y, a.width, a.height, a.color);
    }
  }

  private _inBounds(tx: number, ty: number): boolean {
    return tx >= 0 && ty >= 0 && tx < this.cols && ty < this.rows;
  }
}

File: packages/c/src/world.c
#include "transmute.h"
#include <stdlib.h>
#include <string.h>

#define TM_EMPTY 0
#define TM_SOLID 1
#define TM_MAX_ACTORS 64

struct TmActor {
    int x, y, w, h;
    uint32_t color;
    char name[64];
    TmWorld *world;
};

struct TmWorld {
    int cols, rows, tile;
    int *tiles;
    uint32_t clear_color;
    uint32_t solid_color;
    TmActor actors[TM_MAX_ACTORS];
    int actor_count;
};

TmWorld *tm_world_create(int cols, int rows, int tile) {
    TmWorld *w = calloc(1, sizeof(TmWorld));
    if (!w) return NULL;
    w->cols = cols;
    w->rows = rows;
    w->tile = tile;
    w->tiles = calloc((size_t)cols * (size_t)rows, sizeof(int));
    w->clear_color = tm_pixel(20, 20, 30, 255);
    w->solid_color = tm_pixel(60, 60, 80, 255);
    if (!w->tiles) {
        free(w);
        return NULL;
    }
    return w;
}

void tm_world_destroy(TmWorld *w) {
    if (!w) return;
    free(w->tiles);
    free(w);
}

void tm_world_set_clear(TmWorld *w, uint32_t c) { w->clear_color = c; }
void tm_world_set_solid_color(TmWorld *w, uint32_t c) { w->solid_color = c; }

void tm_world_fill_border(TmWorld *w) {
    for (int x = 0; x < w->cols; x++) {
        tm_world_set_tile(w, x, 0, TM_SOLID);
        tm_world_set_tile(w, x, w->rows - 1, TM_SOLID);
    }
    for (int y = 0; y < w->rows; y++) {
        tm_world_set_tile(w, 0, y, TM_SOLID);
        tm_world_set_tile(w, w->cols - 1, y, TM_SOLID);
    }
}

void tm_world_set_tile(TmWorld *w, int tx, int ty, int tile) {
    if (tx < 0 || ty < 0 || tx >= w->cols || ty >= w->rows) return;
    w->tiles[tx + ty * w->cols] = tile;
}

int tm_world_is_solid(const TmWorld *w, int tx, int ty) {
    if (tx < 0 || ty < 0 || tx >= w->cols || ty >= w->rows) return 1;
    return w->tiles[tx + ty * w->cols] == TM_SOLID;
}

int tm_world_blocks(const TmWorld *w, int x, int y, int ww, int hh) {
    int min_tx = x / w->tile;
    int max_tx = (x + ww - 1) / w->tile;
    int min_ty = y / w->tile;
    int max_ty = (y + hh - 1) / w->tile;
    for (int tx = min_tx; tx <= max_tx; tx++)
        for (int ty = min_ty; ty <= max_ty; ty++)
            if (tm_world_is_solid(w, tx, ty)) return 1;
    return 0;
}

TmActor *tm_world_add_actor(TmWorld *w, int x, int y, int ww, int hh, uint32_t color, const char *name) {
    if (w->actor_count >= TM_MAX_ACTORS) return NULL;
    TmActor *a = &w->actors[w->actor_count++];
    a->x = x;
    a->y = y;
    a->w = ww;
    a->h = hh;
    a->color = color;
    a->world = w;
    if (name) {
        strncpy(a->name, name, sizeof(a->name) - 1);
        a->name[sizeof(a->name) - 1] = 0;
    } else {
        a->name[0] = 0;
    }
    return a;
}

TmActor *tm_world_find_actor(TmWorld *w, const char *name) {
    for (int i = 0; i < w->actor_count; i++)
        if (strcmp(w->actors[i].name, name) == 0) return &w->actors[i];
    return NULL;
}

int tm_actor_try_move(TmActor *a, int dx, int dy) {
    if (!a->world) {
        a->x += dx;
        a->y += dy;
        return 1;
    }
    int nx = a->x + dx, ny = a->y + dy;
    if (tm_world_blocks(a->world, nx, ny, a->w, a->h)) return 0;
    a->x = nx;
    a->y = ny;
    return 1;
}

void tm_world_render(TmWorld *w, TmContext *ctx) {
    tm_context_fill_rect(ctx, 0, 0, ctx->width, ctx->height, w->clear_color);
    for (int ty = 0; ty < w->rows; ty++)
        for (int tx = 0; tx < w->cols; tx++)
            if (tm_world_is_solid(w, tx, ty))
                tm_context_fill_rect(ctx, tx * w->tile, ty * w->tile, w->tile, w->tile, w->solid_color);
    for (int i = 0; i < w->actor_count; i++) {
        TmActor *a = &w->actors[i];
        tm_context_fill_rect(ctx, a->x, a->y, a->w, a->h, a->color);
    }
}

int tm_world_cols(const TmWorld *w) { return w->cols; }
int tm_world_rows(const TmWorld *w) { return w->rows; }
